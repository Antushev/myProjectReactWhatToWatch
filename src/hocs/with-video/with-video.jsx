import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {TypeVideoPlayer} from './../../utils/const.js';

const TIMEOUT_PLAY_VIDEO = 1000;

const withVideo = (Component, typeVideoPlayer) => {
  class WithVideo extends PureComponent {
    constructor(props) {
      super(props);

      this._idTimer = null;

      this._videoRef = React.createRef();

      this.state = {
        isPlaying: false
      };
    }

    render() {
      const video = this._videoRef.current;

      return (
        <Component
          {...this.props}
        >
          <video className="player__video" ref={this._videoRef} muted="muted"/>;
        </Component>
      );
    }

    componentDidMount() {
      const {posterImage, previewVideo, videoMain} = this.props;

      const video = this._videoRef.current;

      video.src = typeVideoPlayer === TypeVideoPlayer.SMALL_VIDEO_PLAYER ?
        previewVideo : videoMain;
      video.poster = `img/${posterImage}`;

      video.onplay = () => {
        this.setState({
          isPlaying: true
        });
      };

      video.onloadeddata = () => {
        this.setState({
          isPlaying: false
        });
      };
    }

    componentDidUpdate() {
      const {isPlaying} = this.props;
      const video = this._videoRef.current;

      if (isPlaying) {
        this._changeActionPlayForPlayers(video);
      } else {
        this._changeActionPauseForPlayers(video);
      }
    }

    componentWillUnmount() {
      const video = this._videoRef.current;

      video.src = ``;
      video.poster = ``;
      video.onplay = null;
      video.onloadeddata = null;
    }

    _changeActionPlayForPlayers(video) {
      switch (typeVideoPlayer) {
        case TypeVideoPlayer.SMALL_VIDEO_PLAYER:
          this._idTimer = setTimeout(() => {
            video.play();
          }, TIMEOUT_PLAY_VIDEO);
          break;
        case TypeVideoPlayer.BIG_VIDEO_PLAYER:
          video.play();
          break;
        default:
          video.play();
      }
    }

    _changeActionPauseForPlayers(video) {
      switch (typeVideoPlayer) {
        case TypeVideoPlayer.SMALL_VIDEO_PLAYER:
          clearTimeout(this._idTimer);
          video.load();
          break;
        case TypeVideoPlayer.BIG_VIDEO_PLAYER:
          video.pause();
          break;
        default:
          clearTimeout(this._idTimer);
          video.load();
      }
    }
  }

  WithVideo.propTypes = {
    isPlaying: PropTypes.bool.isRequired,
    posterImage: PropTypes.string.isRequired,
    previewVideo: PropTypes.string,
    videoMain: PropTypes.string
  };

  return WithVideo;
};

export {withVideo};
