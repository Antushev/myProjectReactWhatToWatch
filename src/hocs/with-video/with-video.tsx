import * as React from 'react';
import {connect} from 'react-redux';
import {Subtract} from 'utility-types';

import {getFilmActive} from './../../reducer/data/selectors';

import {TypeVideoPlayer} from './../../utils/const';

interface State {
  isPlaying: boolean,
  timeElapsed: number,
  progress: number
}

interface InjectingProps {
  videoProgress: number,
  videoTimeElapsed: number,
  onFullScreenClick: () => void
}

const TIMEOUT_PLAY_VIDEO = 1000;
const START_TIME_ELAPSED = 0;
const START_PROGRESS = 1;

const withVideo = (Component, typeVideoPlayer) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithVideo extends React.PureComponent<T, State> {
    private _idTimer: NodeJS.Timeout;
    private _videoRef: any;

    constructor(props) {
      super(props);

      this.state = {
        isPlaying: false,
        timeElapsed: START_TIME_ELAPSED,
        progress: START_PROGRESS
      };

      this._videoRef = React.createRef();

      this._handleFullScreenClick = this._handleFullScreenClick.bind(this);
    }

    render() {
      clearTimeout(this._idTimer);
      const {progress, timeElapsed} = this.state;
      return (
        <Component
          {...this.props}
          videoProgress={progress}
          videoTimeElapsed={timeElapsed}
          onFullScreenClick={this._handleFullScreenClick}
        >
          <video className="player__video" ref={this._videoRef} muted/>;
        </Component>
      );
    }

    componentDidMount() {
      const {filmActive, posterImage, previewVideo} = this.props;

      const video = this._videoRef.current;

      if (typeVideoPlayer === TypeVideoPlayer.SMALL_VIDEO_PLAYER) {
        video.src = previewVideo;
        video.poster = posterImage;
      } else {
        video.src = filmActive.videoMain;
        video.poster = filmActive.posterImage;
      }


      video.onplay = () => {
        this.setState({
          isPlaying: true
        });
      };

      video.onloadeddata = () => {
        this.setState({
          isPlaying: false
        });
      };

      video.ontimeupdate = () => {
        const progress = Math.floor((video.currentTime / video.duration) * 100);
        const timeElapsed = Math.floor(video.duration - video.currentTime);
        this.setState({
          progress,
          timeElapsed
        });
      };
    }

    componentDidUpdate() {
      clearTimeout(this._idTimer);
      const {isPlaying} = this.props;
      const video = this._videoRef.current;

      if (isPlaying) {
        this._changeActionPlayForPlayers(video);
      } else {
        this._changeActionPauseForPlayers(video);
      }
    }

    componentWillUnmount() {
      const video = this._videoRef.current;

      video.src = ``;
      video.poster = ``;
      video.onplay = null;
      video.onloadeddata = null;
      video.ontimeupdate = null;

      if (this._idTimer) {
        clearTimeout(this._idTimer);
      }
    }

    _changeActionPlayForPlayers(video) {
      switch (typeVideoPlayer) {
        case TypeVideoPlayer.SMALL_VIDEO_PLAYER:
          this._idTimer = setTimeout(() => {
            video.play();
          }, TIMEOUT_PLAY_VIDEO);
          break;
        case TypeVideoPlayer.BIG_VIDEO_PLAYER:
          video.play();
          break;
        default:
          video.play();
      }
    }

    _changeActionPauseForPlayers(video) {
      switch (typeVideoPlayer) {
        case TypeVideoPlayer.SMALL_VIDEO_PLAYER:
          clearTimeout(this._idTimer);
          video.load();
          break;
        case TypeVideoPlayer.BIG_VIDEO_PLAYER:
          video.pause();
          break;
        default:
          clearTimeout(this._idTimer);
          video.load();
      }
    }

    _handleFullScreenClick() {
      const video = this._videoRef.current;

      if (!document.fullscreenElement) {
        video.requestFullscreen().catch((err) => err);
      } else {
        document.exitFullscreen().catch((err) => err);
      }
    }
  }

  return WithVideo;
};

const mapStateToProps = (state) => {
  return {
    filmActive: getFilmActive(state),
  };
};

export {withVideo};
export default (Component, typeVideoPlayer) => connect(mapStateToProps, null)(withVideo(Component, typeVideoPlayer));
