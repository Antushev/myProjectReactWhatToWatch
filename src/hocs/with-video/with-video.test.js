import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';

import {withVideo} from './with-video.jsx';

const MockComponent = (props) => {
  const {children} = props;

  return <div>{children}</div>;
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};
const MockComponentWrapped = withVideo(MockComponent);

const posterImage = `src/picture.jpg`;
const previewVideo = `src/video.mp4`;

describe(`Tests HOC with-video`, () => {
  it(`Test HOC with-video`, () => {
    const tree = renderer
      .create(
          <MockComponentWrapped
            isPlaying={false}
            posterImage={posterImage}
            previewVideo={previewVideo}
          />, {
            createNodeMock() {
              return {};
            }
          }
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
