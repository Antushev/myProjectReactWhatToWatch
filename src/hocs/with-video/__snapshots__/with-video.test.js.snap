// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests HOC with-video Test HOC with-video 1`] = `
<div>
  <video
    className="player__video"
    muted="muted"
  />
  ;
</div>
`;
